stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+        sortdb: &SortitionDB,
+    ) -> Result<BlockSnapshot, StacksHttpResponse> {
+        SortitionDB::get_canonical_burn_chain_tip(sortdb.conn()).map_err(|e| {
+            StacksHttpResponse::new_error(
+                &preamble,
+                &HttpServerError::new(format!("Failed to load canonical burnchain tip: {:?}", &e)),
+            )
+        })
+    }
+
+    /// Helper to get the current Stacks epoch
+    fn get_stacks_epoch(
+        &self,
+        preamble: &HttpRequestPreamble,
+        sortdb: &SortitionDB,
+        block_height: u64,
+    ) -> Result<StacksEpoch, StacksHttpResponse> {
+        SortitionDB::get_stacks_epoch(sortdb.conn(), block_height)
+            .map_err(|e| {
+                StacksHttpResponse::new_error(&preamble, &HttpServerError::new(format!("Could not load Stacks epoch for canonical burn height: {:?}", &e)))
+            })?
+            .ok_or_else(|| {
+                let msg = format!(
+                    "Failed to get fee rate estimate because could not load Stacks epoch for canonical burn height = {}",
+                    block_height
+                );
+                warn!("{}", &msg);
+                StacksHttpResponse::new_error(&preamble, &HttpServerError::new(msg))
+            })
+    }
+
+    /// Helper to get the Stacks tip
+    fn get_stacks_chain_tip(
+        &self,
+        preamble: &HttpRequestPreamble,
+        sortdb: &SortitionDB,
+        chainstate: &StacksChainState,
+    ) -> Result<StagingBlock, StacksHttpResponse> {
+        chainstate
+            .get_stacks_chain_tip(sortdb)
+            .map_err(|e| {
+                let msg = format!("Failed to load stacks chain tip: {:?}", &e);
+                warn!("{}", &msg);
+                StacksHttpResponse::new_error(&preamble, &HttpServerError::new(msg))
+            })?
+            .ok_or_else(|| {
+                let msg =
+                    "No stacks tip exists yet. Perhaps no blocks have been processed by this node"
+                        .to_string();
+                warn!("{}", &msg);
+                StacksHttpResponse::new_error(&preamble, &HttpNotFound::new(msg))
+            })
+    }
+}
+
+/// A decoded HttpRequest for use in Stacks
+#[derive(Debug, Clone, PartialEq)]
+pub struct StacksHttpRequest {
+    preamble: HttpRequestPreamble,
+    contents: HttpRequestContents,
+}
+
+impl StacksHttpRequest {
+    pub fn new(preamble: HttpRequestPreamble, contents: HttpRequestContents) -> Self {
+        Self { preamble, contents }
+    }
+
+    /// Instantiate a request to a remote Stacks peer
+    /// `path` is just the request path.  Query arguments are added via `contents`.  Any query
+    /// component to `path` will be silently dropped.
+    ///
+    /// In Stacks, all requests must have a known content length.  If it cannot be calculated, then
+    /// this method will fail.
+    pub fn new_for_peer(
+        peerhost: PeerHost,
+        verb: String,
+        path: String,
```

Comment:
I wonder if this path should take something more like a URL (or equivalent) https://github.com/stacks-network/stacks-blockchain/pull/3947

It will make sure we don't inadvertantly have extra slashes where not wanted. Not by any means a blocking change. but might be worth thinking about.

Reply:
We sanitize `path` in the inner call to `decode_request_path`, which extracts both the normalized path and the query string. The `path` here just captures everything that follows the host.